### DISCUSSION TITLE
How to efficiently incorporate sample projects form GitHub into your solution?

### DISCUSSION THREAD

**Discussion by hasinli66** [Mapsui/Mapsui#1986] [Q&A]
Hi, 

For some time now I’ve been trying to use the NuGet library Mapsui to create a MAUI Windows application.

The instruction are very clear and simple to get a basic app running that will open up OpenStreepMap and allow you to zoom and pan.

Beyond that (say loading SHP files, adding points etc), the documentation just stops (at least in a language I can understand) and you are referred to the sample projects.

The problem is, the sample projects have several dependency projects which refer to dozens of classes with no order I can fathom. When I try and copy all the projects over in the list of dependencies that are required, they fail to build due to more missing references, packages or something else?

Sometimes I figure out that it’s a missing class maybe, but copying that them over just introduces more missing reference warnings.

Am I missing a simple trick to incorporate the sample projects?

Any help, tips or advice would be much appreciated.

**Comment by pauldendulk** [comment]
Your summary is correct. There is a simple getting started and everything else is in the samples. The samples work, but may not be easy to use. Sometimes we need to put more time in it and sometimes we may not be aware of what is hard for our users. Perhaps you can help us by going through a specific case. Do you have an example of what you are trying to do?

Some general remarks that may help.
- You should distinguish: 1) The sample projects that you can run. 2) The individual sample cases which are classes demonstrating a certain case. The cases are classes that are derived from ISample or IMapControlSample sample. The sample projects are mainly intended to show the sample cases not to show how you should build and application.  It is the sample cases that you need to copy into the project that you created with the getting started.
- The samples are dynamically loaded to make it easy for ourselves. This is not what our users should try to do.
- The most common type of samples are the ones that create a Map. They are derived from ISample. If you copy the code of such a sample the ISample interface should be removed.
- The idea is that you should not need to copy that much code. So, if you do need a lot of code something maybe wrong. For instance the samples may use things that are in Mapsui.Tiling and Mapsui.Nts. You need to install those nuget package, not copy the code.
- The samples often use data which is in EmbeddedResources and then deployed at the start. This is for us to make it easy to use these samples on all platforms. There are probably easier ways to load the data on your platform.

**Comment by hasinli66** [comment]
Hi Paul,

Thank you very much for the clarification.

I am a second-year IT student working on a Windows application.
We have found that it is indeed a bit challenging, so we want to limit ourselves to just adding
a couple features, as follows:

  *   When the map opens, it should center and zoom according to the location of the device (longtitude, lattitude),
  *   A blue dot should appear at the user's location,
  *   The user should be able to add a pin at the desired meeting place (similar to searching on Google Maps).
  *   The pins should be visible to other users.


Best regards,
Mustafa

**Reply by pauldendulk** [reply]
With 'Do you have an example of what you are trying to do?' I was asking for a specific sample that you were trying to use in your project, not what you want to do in general.

**Reply by hasinli66** [reply]
I tried to draw some inspiration from this sample:

mapsui / samples / mapsui.samples.common / maps / navigation / CenterOnLocationSample.cs

But I don't understand how I can draw the blue dot on top of the zoomed map.

**Comment by hasinli66** [comment]
P.s. we use the UI framework MAUI and the design pattern MVVM for our project.

________________________________

**Comment by hasinli66** [comment]
I would like to show you my code. I am new into asynchronous programming, working with mapsui and certain new keywords like await, so there might be some mistakes. The code also doesn't still adhere to the mvvm pattern. I want to do that later after making the code work.

I have two methods:

- GetLocation() to request the lattitude and longtitude of the user.
- UpdateMapWithLocation() to zoom the map to an extent based on the location of the user.

using Spark.ViewModels;
using Mapsui.UI.Maui;
using Mapsui.Projections;
using Mapsui.Tiling;
using Mapsui;
using Mapsui.Extensions;
using Mapsui.Layers;
using Mapsui.Providers;

namespace Spark.Views;

public partial class MapPage : ContentPage
{

   private double _latitude;
   private double _longitude;
   private Mapsui.Map map;

   public MapPage(MapViewModel vm)
   {
      InitializeComponent();
      BindingContext = vm;

      UpdateMapWithLocation();
 }

   public void UpdateMapWithLocation()
   {
      GetLocation();

      var mapControl = new MapView();
      mapControl.Map?.Layers.Add(Mapsui.Tiling.OpenStreetMap.CreateTileLayer());

      map = new Mapsui.Map();
      map.Layers.Add(OpenStreetMap.CreateTileLayer());

      var centerPoint = new MPoint(_longitude, _latitude); // Swap longitude and latitude here
      var sphericalMercatorCoordinate = SphericalMercator.FromLonLat(centerPoint.X, centerPoint.Y).ToMPoint();
      map.Home = n => n.CenterOnAndZoomTo(sphericalMercatorCoordinate, n.Resolutions[14]);

      Content = mapControl;
   }

   private async void GetLocation()
   {
      try
      {
         var request = new GeolocationRequest(GeolocationAccuracy.Best);
         var location = await Geolocation.GetLocationAsync(request);

         _latitude = location.Latitude;
         _longitude = location.Longitude;

      }
      catch (PermissionException ex)
      {
         var label = new Label();
         label.Text = "Unable to determine location";
         Content = label;
      }
   }
}
